### PR TITLE
add phonetic transcriptions term meta type

### PR DESCRIPTION
**Pull request by StefanVukovic99**
part 2 of #422. Adds a 3rd term meta type for IPA transcriptions.

The types could maybe be improved somehow, there is a ts error in `anki-note-data-creator.js` because of the union type. 

Also, I missed 2 references to `dictionary-data-util.js` in #429, should i fix that in this pr or make another?

### CODE REVIEW

**Comment by toasted-nutbread** [comment]
> Also, I missed 2 references to dictionary-data-util.js in https://github.com/themoeway/yomitan/pull/429, should i fix that in this pr or make another?

Are you referring to the JSON config files? It's fine to be in this since it's close enough related I'd say.

File: ext/js/data/sandbox/anki-note-data-creator.js [hunk 1]
```diff
@@ -353,6 +385,7 @@ export class AnkiNoteDataCreator {
         const expressions = this.createCachedValue(this._getTermExpressions.bind(this, dictionaryEntry));
         const frequencies = this.createCachedValue(this._getTermFrequencies.bind(this, dictionaryEntry));
         const pitches = this.createCachedValue(this._getTermPitches.bind(this, dictionaryEntry));
+        const transcriptions = this.createCachedValue(this._getTermPhoneticTranscriptions.bind(this, dictionaryEntry));
```
Comment (toasted-nutbread) [inline ext/js/data/sandbox/anki-note-data-creator.js:388]: This is named just `transcriptions` while `phoneticTranscriptions` is used above. Should `phoneticTranscriptions` be used for both?

File: types/ext/dictionary-database.d.ts [hunk 2]
```diff
@@ -118,12 +118,19 @@ export type DatabaseTermMetaPitch = {
     dictionary: string;
 };
 
+export type DatabaseTermMetaIpa = {
```
Comment (toasted-nutbread) [inline types/ext/dictionary-database.d.ts:121]: Naming suggestion: Instead of using `ipa` and variants in type names, use `PhoneticTranscription`. The `'ipa'` is good for the dictionary data schema, since shortening there is better for overall file/database size.

File: types/ext/dictionary-database.d.ts [hunk 3]
```diff
@@ -143,6 +150,15 @@ export type TermMetaPitch = {
     dictionary: string;
 };
 
+export type TermMetaIPA = {
+    index: number;
+    term: string;
+    mode: 'ipa';
```
Comment (toasted-nutbread) [inline types/ext/dictionary-database.d.ts:156]: Put the type disambiguation field as the first field. (`mode: 'ipa';`)

File: ext/js/dictionary/dictionary-data-util.js [hunk 4]
```diff
@@ -146,23 +146,20 @@ export class DictionaryDataUtil {
...
                 dictionaryGroupedPronunciationList = [];
                 groupedPronunciationsMap.set(dictionary, dictionaryGroupedPronunciationList);
             }
-            for (const {position, nasalPositions, devoicePositions, tags} of pitches) {
-                let groupedPronunciation = this._findExistingGroupedPronunciation(reading, position, nasalPositions, devoicePositions, tags, dictionaryGroupedPronunciationList);
+            for (const pronunciation of pronunciations) {
+                let groupedPronunciation = this._findExistingGroupedPronunciation(reading, pronunciation, dictionaryGroupedPronunciationList);
                 if (groupedPronunciation === null) {
                     groupedPronunciation = {
+                        ...pronunciation,
```
Comment (toasted-nutbread) [inline ext/js/dictionary/dictionary-data-util.js:160]: Can this be made to not be a rest expansion? I.e. just `pronunciation`?

File: ext/js/dictionary/dictionary-data-util.js [hunk 5]
```diff
@@ -288,23 +304,28 @@ export class DictionaryDataUtil {
...
-                this._areTagListsEqual(pitchInfo.tags, tags)
+                groupedPronunciation.reading === reading &&
+                this._areTagListsEqual(groupedPronunciation.tags, pronunciation.tags) &&
+                (
+                    pronunciation.type === 'pitch-accent' &&
+                    groupedPronunciation.type === 'pitch-accent' &&
+                    groupedPronunciation.position === pronunciation.position &&
+                    this._areArraysEqual(groupedPronunciation.nasalPositions, pronunciation.nasalPositions) &&
+                    this._areArraysEqual(groupedPronunciation.devoicePositions, pronunciation.devoicePositions)
+                    ||
```
Comment (toasted-nutbread) [inline ext/js/dictionary/dictionary-data-util.js:322]: && and || precedence combination unclear and somewhat inefficient, create a helper function for this and do a `switch` on `.type`.

File: ext/js/display/display-generator.js [hunk 6]
```diff
@@ -641,35 +641,61 @@ export class DisplayGenerator {
      */
     _createPronunciation(details) {
         const jp = this._japaneseUtil;
-        const {reading, position, nasalPositions, devoicePositions, tags, exclusiveTerms, exclusiveReadings} = details;
-        const morae = jp.getKanaMorae(reading);
 
-        const node = this._instantiate('pronunciation');
+        switch (details.type) {
+            case 'pitch-accent': {
```
Comment (toasted-nutbread) [inline ext/js/display/display-generator.js:646]: For each of the switch cases, create a sub-function to handle that case.
```js
case 'pitch-accent': return this._createPronunciationPitchAccent(details); break;
case 'phonetic-transcription': return this._createPronunciationPhoneticTranscription(details); break;
```

**Review by toasted-nutbread — CHANGES_REQUESTED** [review]
Left some initial thoughts.

File: ext/js/data/sandbox/anki-note-data-creator.js [hunk 7]
```diff
@@ -485,15 +519,17 @@ export class AnkiNoteDataCreator {
...
     _getTermPitches(dictionaryEntry) {
         // eslint-disable-next-line @typescript-eslint/no-this-alias
         const self = this;
         const results = [];
         const {headwords} = dictionaryEntry;
-        for (const {headwordIndex, dictionary, dictionaryIndex, dictionaryPriority, pitches} of dictionaryEntry.pronunciations) {
+        for (const {headwordIndex, dictionary, dictionaryIndex, dictionaryPriority, pronunciations} of dictionaryEntry.pronunciations) {
             const {term, reading} = headwords[headwordIndex];
+            /** @type {import('dictionary').PitchAccent[]} */
+            const pitches = pronunciations.filter((v) => v.type === 'pitch-accent');
```
Comment (toasted-nutbread) [inline ext/js/data/sandbox/anki-note-data-creator.js:532]: To fix this, just do like this
```js
            /** @type {import('dictionary').PitchAccent[]} */
            const pitches = [];
            for (const pronunciation of pronunciations) {
                if (pronunciation.type !== 'pitch-accent') { continue; }
                pitches.push(pronunciation);
            }
```

Comment (toasted-nutbread) [reply ext/js/data/sandbox/anki-note-data-creator.js:532]: Same for other locations. Create a utility function for doing this. This can probably be templatized, I'll look into how to easily do that.

File: types/ext/dictionary.d.ts [hunk 8]
```diff
@@ -365,15 +365,18 @@ export type TermPronunciation = {
...
 };
 
+export type Pronunciation = PitchAccent | IpaTranscription;
+
 /**
  * Pitch accent information for a term, represented as the position of the downstep.
  */
-export type TermPitch = {
+export type PitchAccent = {
+    type: 'pitch-accent';
```
Comment (toasted-nutbread) [inline types/ext/dictionary.d.ts:379]: Add comment describing that this is used for type disambiguation.

File: ext/js/data/sandbox/anki-note-data-creator.js (hunk 7 above)
Comment (toasted-nutbread) [reply ext/js/data/sandbox/anki-note-data-creator.js:532]:
```js
// ext/js/data/sandbox/anki-note-data-creator.js

// No type annotation necessary before declaration
const pitches = this._getPronunciationsOfType(pronunciations, 'pitch-accent');


// ext/js/data/sandbox/anki-note-data-creator.js

/**
 * @template {import('dictionary').PronunciationType} T
 * @param {import('dictionary').Pronunciation[]} pronunciations
 * @param {T} type
 * @returns {import('dictionary').PronunciationGeneric<T>[]}
 */
_getPronunciationsOfType(pronunciations, type) {
    /** @type {import('dictionary').PronunciationGeneric<T>[]} */
    const results = [];
    for (const pronunciation of pronunciations) {
        if (pronunciation.type === type) { continue; }
        // This is type safe, but for some reason the cast is needed.
        results.push(/** @type {import('dictionary').PronunciationGeneric<T>} */ (pronunciation));
    }
    return results;
}


// types/ext/dictionary.d.ts

export type PronunciationType = Pronunciation['type'];

export type PronunciationGeneric<T extends PronunciationType> = Extract<Pronunciation, {type: T}>;
```

File: ext/js/dictionary/dictionary-data-util.js (hunk 4 above)
Comment (StefanVukovic99) [reply ext/js/dictionary/dictionary-data-util.js:160]: Not sure what you mean, like this?
```
export type GroupedPronunciation = {
    pronunciation: GroupedPitchAccentInternal | GroupedPhoneticTranscriptionInternal;
    exclusiveTerms: string[];
    exclusiveReadings: string[];
};
```

Comment (toasted-nutbread) [reply ext/js/dictionary/dictionary-data-util.js:160]: Correct. Expanding it with `...` seems like it should be unnecessary and would just create a more complex object to construct.

**Comment by toasted-nutbread** [comment]
Assuming you're based off latest master, you can run `npm run test-code-write` to fix the test data.

File: ext/js/dictionary/dictionary-data-util.js (hunk 4 above)
Comment (StefanVukovic99) [reply ext/js/dictionary/dictionary-data-util.js:160]: Sorry, I can't get the types to work when it's like that.

Comment (toasted-nutbread) [reply ext/js/dictionary/dictionary-data-util.js:160]:
```patch
diff --git a/ext/js/data/sandbox/anki-note-data-creator.js b/ext/js/data/sandbox/anki-note-data-creator.js
index e4e0b64d..f27cee86 100644
--- a/ext/js/data/sandbox/anki-note-data-creator.js
+++ b/ext/js/data/sandbox/anki-note-data-creator.js
@@ -196,9 +196,11 @@ export class AnkiNoteDataCreator {
             for (const {dictionary, pronunciations} of DictionaryDataUtil.getGroupedPronunciations(dictionaryEntry)) {
                 /** @type {import('anki-templates').Pitch[]} */
                 const pitches = [];
-                for (const pronunciation of pronunciations) {
+                for (const groupedPronunciation of pronunciations) {
+                    const {pronunciation} = groupedPronunciation;
                     if (pronunciation.type !== 'pitch-accent') { continue; }
-                    const {terms, reading, position, nasalPositions, devoicePositions, tags, exclusiveTerms, exclusiveReadings} = pronunciation;
+                    const {terms, reading, exclusiveTerms, exclusiveReadings} = groupedPronunciation;
+                    const {position, nasalPositions, devoicePositions, tags} = pronunciation;
                     pitches.push({
                         expressions: terms,
                         reading,
@@ -225,9 +227,11 @@ export class AnkiNoteDataCreator {
         if (dictionaryEntry.type === 'term') {
             for (const {dictionary, pronunciations} of DictionaryDataUtil.getGroupedPronunciations(dictionaryEntry)) {
                 const phoneticTranscriptions = [];
-                for (const pronunciation of pronunciations) {
+                for (const groupedPronunciation of pronunciations) {
+                    const {pronunciation} = groupedPronunciation;
                     if (pronunciation.type !== 'phonetic-transcription') { continue; }
-                    const {terms, reading, ipa, tags, exclusiveTerms, exclusiveReadings} = pronunciation;
+                    const {ipa, tags} = pronunciation;
+                    const {terms, reading, exclusiveTerms, exclusiveReadings} = groupedPronunciation;
                     phoneticTranscriptions.push({
                         expressions: terms,
                         reading,
diff --git a/ext/js/dictionary/dictionary-data-util.js b/ext/js/dictionary/dictionary-data-util.js
index f889b8d0..408b1975 100644
--- a/ext/js/dictionary/dictionary-data-util.js
+++ b/ext/js/dictionary/dictionary-data-util.js
@@ -157,7 +157,7 @@ export class DictionaryDataUtil {
                 let groupedPronunciation = this._findExistingGroupedPronunciation(reading, pronunciation, dictionaryGroupedPronunciationList);
                 if (groupedPronunciation === null) {
                     groupedPronunciation = {
-                        ...pronunciation,
+                        pronunciation,
                         terms: new Set(),
                         reading
                     };
@@ -174,41 +174,19 @@ export class DictionaryDataUtil {
             /** @type {import('dictionary-data-util').GroupedPronunciation[]} */
             const pronunciations2 = [];
             for (const groupedPronunciation of dictionaryGroupedPronunciationList) {
-                const {terms, reading} = groupedPronunciation;
+                const {pronunciation, terms, reading} = groupedPronunciation;
                 const exclusiveTerms = !this._areSetsEqual(terms, allTerms) ? this._getSetIntersection(terms, allTerms) : [];
                 const exclusiveReadings = [];
                 if (multipleReadings) {
                     exclusiveReadings.push(reading);
                 }
-                switch (groupedPronunciation.type) {
-                    case 'pitch-accent': {
-                        const {position, nasalPositions, devoicePositions, tags} = groupedPronunciation;
-                        pronunciations2.push({
-                            type: 'pitch-accent',
-                            terms: [...terms],
-                            reading,
-                            position,
-                            nasalPositions,
-                            devoicePositions,
-                            tags,
-                            exclusiveTerms,
-                            exclusiveReadings
-                        });
-                        break;
-                    } case 'phonetic-transcription': {
-                        const {ipa, tags} = groupedPronunciation;
-                        pronunciations2.push({
-                            type: 'phonetic-transcription',
-                            terms: [...terms],
-                            reading,
-                            ipa,
-                            tags,
-                            exclusiveTerms,
-                            exclusiveReadings
-                        });
-                        break;
-                    }
-                }
+                pronunciations2.push({
+                    pronunciation,
+                    terms: [...terms],
+                    reading,
+                    exclusiveTerms,
+                    exclusiveReadings
+                });
             }
             results2.push({dictionary, pronunciations: pronunciations2});
         }
@@ -339,18 +317,19 @@ export class DictionaryDataUtil {
      * @returns {boolean}
      */
     static _arePronunciationsEquivalent(groupedPronunciation, pronunciation) {
-        if (!this._areTagListsEqual(groupedPronunciation.tags, pronunciation.tags)) {
+        const pronunciation0 = groupedPronunciation.pronunciation;
+        if (!this._areTagListsEqual(pronunciation0.tags, pronunciation.tags)) {
             return false;
         }
 
-        if (pronunciation.type === 'pitch-accent' && groupedPronunciation.type === 'pitch-accent') {
-            if (groupedPronunciation.position !== pronunciation.position ||
-                !this._areArraysEqual(groupedPronunciation.nasalPositions, pronunciation.nasalPositions) ||
-                !this._areArraysEqual(groupedPronunciation.devoicePositions, pronunciation.devoicePositions)) {
+        if (pronunciation.type === 'pitch-accent' && pronunciation0.type === 'pitch-accent') {
+            if (pronunciation0.position !== pronunciation.position ||
+                !this._areArraysEqual(pronunciation0.nasalPositions, pronunciation.nasalPositions) ||
+                !this._areArraysEqual(pronunciation0.devoicePositions, pronunciation.devoicePositions)) {
                 return false;
             }
-        } else if (pronunciation.type === 'phonetic-transcription' && groupedPronunciation.type === 'phonetic-transcription') {
-            if (groupedPronunciation.ipa !== pronunciation.ipa) {
+        } else if (pronunciation.type === 'phonetic-transcription' && pronunciation0.type === 'phonetic-transcription') {
+            if (pronunciation0.ipa !== pronunciation.ipa) {
                 return false;
             }
         } else {
diff --git a/ext/js/display/display-generator.js b/ext/js/display/display-generator.js
index 393092a5..d3f77883 100644
--- a/ext/js/display/display-generator.js
+++ b/ext/js/display/display-generator.js
@@ -621,7 +621,7 @@ export class DisplayGenerator {
         n1.appendChild(tag);
 
         let hasTags = false;
-        for (const {tags} of pronunciations) {
+        for (const {pronunciation: {tags}} of pronunciations) {
             if (tags.length > 0) {
                 hasTags = true;
                 break;
@@ -640,21 +640,24 @@ export class DisplayGenerator {
      * @returns {HTMLElement}
      */
     _createPronunciation(details) {
-        switch (details.type) {
+        const {pronunciation} = details;
+        switch (pronunciation.type) {
             case 'pitch-accent':
-                return this._createPronunciationPitchAccent(details);
+                return this._createPronunciationPitchAccent(pronunciation, details);
             case 'phonetic-transcription':
-                return this._createPronunciationPhoneticTranscription(details);
+                return this._createPronunciationPhoneticTranscription(pronunciation, details);
         }
     }
 
 
     /**
-     * @param {import('dictionary-data-util').GroupedPhoneticTranscription} details
+     * @param {import('dictionary').PhoneticTranscription} phoneticTranscription
+     * @param {import('dictionary-data-util').GroupedPronunciation} groupedPronunciation
      * @returns {HTMLElement}
      */
-    _createPronunciationPhoneticTranscription(details) {
-        const {ipa, tags, exclusiveTerms, exclusiveReadings} = details;
+    _createPronunciationPhoneticTranscription(phoneticTranscription, groupedPronunciation) {
+        const {ipa, tags} = phoneticTranscription;
+        const {exclusiveTerms, exclusiveReadings} = groupedPronunciation;
 
         const node = this._instantiate('pronunciation');
 
@@ -674,13 +677,15 @@ export class DisplayGenerator {
     }
 
     /**
-     * @param {import('dictionary-data-util').GroupedPitchAccent} details
+     * @param {import('dictionary').PitchAccent} pitchAccent
+     * @param {import('dictionary-data-util').GroupedPronunciation} groupedPronunciation
      * @returns {HTMLElement}
      */
-    _createPronunciationPitchAccent(details) {
+    _createPronunciationPitchAccent(pitchAccent, groupedPronunciation) {
         const jp = this._japaneseUtil;
 
-        const {reading, position, nasalPositions, devoicePositions, tags, exclusiveTerms, exclusiveReadings} = details;
+        const {position, nasalPositions, devoicePositions, tags} = pitchAccent;
+        const {reading, exclusiveTerms, exclusiveReadings} = groupedPronunciation;
         const morae = jp.getKanaMorae(reading);
 
         const node = this._instantiate('pronunciation');
diff --git a/types/ext/dictionary-data-util.d.ts b/types/ext/dictionary-data-util.d.ts
index f952da62..4ab06f11 100644
--- a/types/ext/dictionary-data-util.d.ts
+++ b/types/ext/dictionary-data-util.d.ts
@@ -63,51 +63,20 @@ export type KanjiFrequency = {
 
 export type TermFrequencyType = 'popular' | 'rare' | 'normal';
 
-export type GroupedPitchAccentInternal = {
-    type: 'pitch-accent';
+export type GroupedPronunciationInternal = {
+    pronunciation: Dictionary.Pronunciation;
     terms: Set<string>;
     reading: string;
-    position: number;
-    nasalPositions: number[];
-    devoicePositions: number[];
-    tags: Dictionary.Tag[];
 };
 
-export type GroupedPhoneticTranscriptionInternal = {
-    type: 'phonetic-transcription';
-    terms: Set<string>;
-    reading: string;
-    tags: Dictionary.Tag[];
-    ipa: string;
-};
-
-export type GroupedPronunciationInternal = GroupedPitchAccentInternal | GroupedPhoneticTranscriptionInternal;
-
-export type GroupedPitchAccent = {
-    type: 'pitch-accent';
+export type GroupedPronunciation = {
+    pronunciation: Dictionary.Pronunciation;
     terms: string[];
     reading: string;
-    position: number;
-    nasalPositions: number[];
-    devoicePositions: number[];
-    tags: Dictionary.Tag[];
     exclusiveTerms: string[];
     exclusiveReadings: string[];
 };
 
-export type GroupedPhoneticTranscription = {
-    type: 'phonetic-transcription';
-    terms: string[];
-    reading: string;
-    tags: Dictionary.Tag[];
-    ipa: string;
-    exclusiveTerms: string[];
-    exclusiveReadings: string[];
-};
-
-
-export type GroupedPronunciation = GroupedPitchAccent | GroupedPhoneticTranscription;
-
 export type DictionaryGroupedPronunciations = {
     dictionary: string;
     pronunciations: GroupedPronunciation[];
```

File: ext/data/templates/default-anki-field-templates.handlebars [hunk 9]
```diff
@@ -229,6 +229,25 @@
 {{/inline}}
 {{! End Pitch Accents }}
 
+{{#*inline "phonetic-transcriptions"}}
+    <ul>
```
Comment (toasted-nutbread) [inline ext/data/templates/default-anki-field-templates.handlebars:233]: If empty, this generates `<ul></ul>`. Generating an empty value would be preferred, similar to the definition of `"frequencies"`.

File: ext/js/dictionary/dictionary-data-util.js [hunk 10]
```diff
@@ -288,26 +321,43 @@ export class DictionaryDataUtil {
...
+     * @param {import('dictionary-data-util').GroupedPronunciationInternal} groupedPronunciation
+     * @param {import('dictionary').Pronunciation} pronunciation
+     * @returns {boolean}
+     */
+    static _arePronunciationsEquivalent(groupedPronunciation, pronunciation) {
+        if (!this._areTagListsEqual(groupedPronunciation.tags, pronunciation.tags)) {
+            return false;
+        }
+
+        if (pronunciation.type === 'pitch-accent' && groupedPronunciation.type === 'pitch-accent') {
```
Comment (toasted-nutbread) [inline ext/js/dictionary/dictionary-data-util.js:346]: This can be simplified. Put the type check at the start of the function; that check is faster, and if those aren't the same, nothing else is.

`if (p.type !== gp.type) { return false; }`

Then you can `switch` on type.

Comment (StefanVukovic99) [reply ext/js/dictionary/dictionary-data-util.js]: I tried that first, but didn't like the redundant type check in each case to get it to calm down about the types:

```
static _arePronunciationsEquivalent({pronunciation: pronunciation1}, pronunciation2) {
        if (pronunciation1.type !== pronunciation2.type) {
            return false;
        }
        if (!this._areTagListsEqual(pronunciation1.tags, pronunciation2.tags)) {
            return false;
        }
        switch (pronunciation1.type) {
            case 'pitch-accent':
                if (
                    pronunciation2.type === 'pitch-accent' &&
                    (pronunciation1.position !== pronunciation2.position ||
                     !this._areArraysEqual(pronunciation1.nasalPositions, pronunciation2.nasalPositions) ||
                     !this._areArraysEqual(pronunciation1.devoicePositions, pronunciation2.devoicePositions))
                ) {
                    return false;
                }
                break;
            case 'phonetic-transcription':
                if (
                    pronunciation2.type === 'phonetic-transcription' &&
                    pronunciation1.ipa !== pronunciation2.ipa
                ) {
                    return false;
                }
                break;
        }

        return true;
    }
```

Is that still preferable?

Comment (toasted-nutbread) [reply ext/js/dictionary/dictionary-data-util.js]: Oh that's annoying that TypeScript can't infer that. Here's the simplest it can probably be made:

```js
static _arePronunciationsEquivalent({pronunciation: pronunciation1}, pronunciation2) {
    if (
        pronunciation1.type !== pronunciation2.type ||
        !this._areTagListsEqual(pronunciation1.tags, pronunciation2.tags)
    ) {
        return false;
    }
    switch (pronunciation1.type) {
        case 'pitch-accent':
        {
            // This cast is valid based on the type check at the start of the function.
            const pitchAccent2 = /** @type {import('dictionary').PitchAccent} */ (pronunciation2);
            return (
                pronunciation1.position === pitchAccent2.position &&
                this._areArraysEqual(pronunciation1.nasalPositions, pitchAccent2.nasalPositions) &&
                this._areArraysEqual(pronunciation1.devoicePositions, pitchAccent2.devoicePositions)
            );
        }
        case 'phonetic-transcription':
        {
            // This cast is valid based on the type check at the start of the function.
            const phoneticTranscription2 = /** @type {import('dictionary').PhoneticTranscription} */ (pronunciation2);
            return pronunciation1.ipa === phoneticTranscription2.ipa;
        }
    }
    return true;
}
```

**Comment by StefanVukovic99** [comment]
> Assuming you're based off latest master, you can run `npm run test-code-write` to fix the test data.

```shell
npm run test-code-write

> yomitan@0.0.0 test-code-write
> vitest run --config test/data/vitest.write.config.json


 RUN  v0.34.6 /media/stefan/DATA/EDesktop/Projekti/Coding/yezichak

 ❯ test/dictionary-data.write.js (0)

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Suites 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  test/dictionary-data.write.js [ test/dictionary-data.write.js ]
ReferenceError: TransformStream is not defined
 ❯ node_modules/@zip.js/zip.js/lib/core/constants.js:238:19
 ❯ ext/js/dictionary/dictionary-importer.js:2:31
      1| /*
      2|  * Copyright (C) 2023  Yomitan Authors
       |                               ^
      3|  * Copyright (C) 2020-2022  Yomichan Authors
      4|  *

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯

 Test Files  1 failed (1)
      Tests  no tests
   Start at  22:57:17
   Duration  2.15s (transform 626ms, setup 0ms, collect 0ms, tests 0ms, environment 0ms, prepare 211ms)
```

**Comment by toasted-nutbread** [comment]
Try rebuilding libraries: `npm run build-libs`

**Comment by StefanVukovic99** [comment]
> Try rebuilding libraries: `npm run build-libs`

No change.

**Comment by StefanVukovic99** [comment]
> > Try rebuilding libraries: `npm run build-libs`
> 
> No change.

updated Node to v20, manually deleted node_modules, npm install - now it works

**Comment by toasted-nutbread** [comment]
🤔 

```log
~\yomitan>npm ci
npm WARN deprecated @types/handlebars@4.1.0: This is a stub types definition. handlebars provides its own type definitions, so you do not need this installed.      
npm WARN deprecated source-map-resolve@0.6.0: See https://github.com/lydell/source-map-resolve#deprecated

added 492 packages, and audited 493 packages in 8s

108 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

~\yomitan>npm run build-libs

> yomitan@0.0.0 build-libs
> node ./dev/bin/build-libs.js


~\yomitan>npm run test-code-write

> yomitan@0.0.0 test-code-write
> vitest run --config test/data/vitest.write.config.json


 RUN  v0.34.6 ~/yomitan

 ✓ test/dictionary-data.write.js (1) 1386ms
   ✓ Write dictionary data expected data 1385ms

 Test Files  1 passed (1)
      Tests  1 passed (1)
   Start at  17:10:41
   Duration  4.80s (transform 543ms, setup 0ms, collect 830ms, tests 1.39s, environment 0ms, prepare 96ms)


~\yomitan>git diff --shortstat
 3 files changed, 893 insertions(+), 31 deletions(-)
```

**Review by djahandarie — DISMISSED** [review]

File: .gitignore [hunk 11]
```diff
@@ -5,6 +5,7 @@ node_modules/
 builds/
 
 dictionaries/
+dict/
```
Comment (toasted-nutbread) [inline .gitignore:8]: Why is this added?

Comment (StefanVukovic99) [reply .gitignore]: didn't mean to commit that, i locally have some dict converters in that folder

**Comment by toasted-nutbread** [comment]
I think overall this looks good. The only thing that I think would be good to add is a/some test data in `test/data/translator-test-inputs.json` and `test/data/dictionaries/valid-dictionary1/term_meta_bank_1.json`.

**Comment by StefanVukovic99** [comment]
I've added some test data.

File: package-lock.json [hunk 12]
```diff
@@ -54,6 +54,9 @@
                 "ts-json-schema-generator": "^1.5.0",
                 "typescript": "5.3.3",
                 "vitest": "^0.34.6"
+            },
+            "engines": {
```
Comment (toasted-nutbread) [inline package-lock.json:58]: Oh whoops, guess I forgot to regenerate package-lock in #442.

**Review by toasted-nutbread — APPROVED** [review]

**Review by djahandarie — APPROVED** [review]
Thanks so much both of you for polishing this PR 🙏🙏
